### PR TITLE
Ingestion: Use the same appender for regular samples and reporting (`up` and friends) #6950

### PR DESCRIPTION
**Pull request by liguozhong**
Ingestion: Use the same appender for regular samples and reporting (`up` and friends) #6950

### CODE REVIEW

File: scrape/scrape.go [hunk 1]
```diff
@@ -1275,16 +1283,17 @@ const (
...
+func (sl *scrapeLoop) report(app storage.Appender, start time.Time, duration time.Duration, scraped, appended, seriesAdded int, scrapeErr error) (err error) {
 	sl.scraper.Report(start, duration, scrapeErr)
 
 	ts := timestamp.FromTime(start)
 
 	var health float64
 	if scrapeErr == nil {
 		health = 1
+	} else {
+		app = sl.appender()
```
Comment (codesome) [inline scrape/scrape.go:1295]: Why is it in this `else`? Shouldn't we check for `app == nil`?

Comment (liguozhong) [reply scrape/scrape.go:1295]: When the scrape fails, the app is still not nil.
As mentioned in issue, use a new appender when the scraper fails.

File: scrape/scrape.go [hunk 2]
```diff
@@ -967,12 +967,13 @@ mainLoop:
...
 		// we still call sl.append to trigger stale markers.
-		total, added, seriesAdded, appErr := sl.append(b, contentType, start)
+		scrapeAppender := sl.appender()
+		total, added, seriesAdded, appErr := sl.append(scrapeAppender, b, contentType, start)
 		if appErr != nil {
 			level.Debug(sl.l).Log("msg", "append failed", "err", appErr)
 			// The append failed, probably due to a parse error or sample limit.
 			// Call sl.append again with an empty scrape to trigger stale markers.
-			if _, _, _, err := sl.append([]byte{}, "", start); err != nil {
+			if _, _, _, err := sl.append(scrapeAppender, []byte{}, "", start); err != nil {
```
Comment (brian-brazil) [inline scrape/scrape.go:976]: Don't we need a new appender here?

File: scrape/scrape.go (hunk 1 above)
Comment (brian-brazil) [reply scrape/scrape.go:1295]: Spreading the logic across functions is fragile, I think this should all be in `run`

File: scrape/scrape.go (hunk 2 above)
Comment (liguozhong) [reply scrape/scrape.go]: thanks .done

File: scrape/scrape.go (hunk 1 above)
Comment (liguozhong) [reply scrape/scrape.go]: > Spreading the logic across functions is fragile, I think this should all be in `run`
thanks .done.

image
image

File: scrape/scrape.go [hunk 3]
```diff
@@ -967,12 +967,13 @@ mainLoop:
...
 		// we still call sl.append to trigger stale markers.
-		total, added, seriesAdded, appErr := sl.append(b, contentType, start)
+		scrapeAppender := sl.appender()
+		total, added, seriesAdded, appErr := sl.append(scrapeAppender, b, contentType, start)
 		if appErr != nil {
 			level.Debug(sl.l).Log("msg", "append failed", "err", appErr)
 			// The append failed, probably due to a parse error or sample limit.
 			// Call sl.append again with an empty scrape to trigger stale markers.
-			if _, _, _, err := sl.append([]byte{}, "", start); err != nil {
+			if _, _, _, err := sl.append(sl.appender(), []byte{}, "", start); err != nil {
```
Comment (roidelapluie) [inline scrape/scrape.go:976]: why not reusing  scrapeAppender ?

File: scrape/scrape.go [hunk 4]
```diff
@@ -1045,7 +1048,7 @@ func (sl *scrapeLoop) endOfRunStaleness(last time.Time, ticker *time.Ticker, int
 	// Call sl.append again with an empty scrape to trigger stale markers.
 	// If the target has since been recreated and scraped, the
 	// stale markers will be out of order and ignored.
-	if _, _, _, err := sl.append([]byte{}, "", staleTime); err != nil {
+	if _, _, _, err := sl.append(nil, []byte{}, "", staleTime); err != nil {
 		level.Error(sl.l).Log("msg", "stale append failed", "err", err)
 	}
 	if err := sl.reportStale(staleTime); err != nil {
```
Comment (roidelapluie) [inline scrape/scrape.go:1054]: this is also using an appender

**Comment by roidelapluie** [comment]
I think this PR is not ready t be merged. By reading the code, it looks like we could Commit or Rollback twice.

File: scrape/scrape.go (hunk 3 above)
Comment (liguozhong) [reply scrape/scrape.go]: thanks for your review ,done

File: scrape/scrape.go (hunk 4 above)
Comment (liguozhong) [reply scrape/scrape.go:1054]: thanks for your review ,done

File: scrape/scrape.go (hunk 2 above)
Comment (brian-brazil) [reply scrape/scrape.go:976]: We still need a new appender here, and the old one rolled back.

File: scrape/scrape.go [hunk 5]
```diff
@@ -1284,7 +1294,9 @@ func (sl *scrapeLoop) report(start time.Time, duration time.Duration, scraped, a
 	if scrapeErr == nil {
 		health = 1
 	}
-	app := sl.appender()
+	if app == nil {
```
Comment (brian-brazil) [inline scrape/scrape.go:1297]: It should always be passed in.

Comment (liguozhong) [reply scrape/scrape.go]: thanks.done

**Comment by roidelapluie** [comment]
Hello,

This PR has gone stale for 4 months and I actually forgot about it, so I ended up implementing it in a different way in #7561.

Sorry about that, and thanks for the work you've done.

**Pull request closed without merging**